### PR TITLE
MuZero Reanalyze, regularized policy for MCTS

### PR DESCRIPTION
**Pull request by emailweixu**
Also improved memory usage for MCTS

### CODE REVIEW

File: .ci-cd/build.sh [hunk 1]
```diff
@@ -41,7 +41,8 @@ function check_style() {
 
 function test() {
     cd alf
-    pip3 install --user -e ./nest/cnest
+    pip3 install -e ./nest/cnest
+    export PYTHONPATH=$PYTHONPATH:./nest/cnest
```
Comment (hnyu) [inline .ci-cd/build.sh]: prepending PYTHONPATH to the beginning of the python3 command line is a better choice. Multiple runs of build.sh won't increase the length of PYTHONPATH.

File: alf/algorithms/data_transformer.py [hunk 2]
```diff
@@ -547,6 +548,23 @@ def _transform(self, timestep_or_exp):
             reward=timestep_or_exp.reward.clamp(*self._minmax))
 
 
+@gin.configurable
+class RewardNormalizer(SimpleDataTransformer):
+    def __init__(self, observation_spec, normalizer=None, clip_value=-1.0):
```
Comment (hnyu) [inline alf/algorithms/data_transformer.py]: Why do we need observation_spec here?

File: alf/algorithms/mcts_algorithm.py [hunk 3]
```diff
@@ -204,6 +230,15 @@ def __init__(
                 the games. And it is expected to return a float vector of the same
                 shape as ``steps``.
             known_value_bounds (tuple|None): known bound of the values.
+            unexpanded_value_score (float|str): The value score for an unexpanded
+                child. If 'max'/'min'/'mean', will use the maximum/minimum/mean
+                of the value scores of the expanded siblings. If 'none', when
+                exploration policy is used, will keep the policy for the unexpanded
+                children same as prior; when exporation is not used, 'none' behaves
+                same as 'min'.
+            act_with_exploratin_policy (bool):
```
Comment (hnyu) [inline alf/algorithms/mcts_algorithm.py]: explanation for these three boolean arguments?

File: alf/algorithms/mcts_algorithm.py [hunk 4]
```diff
@@ -325,103 +384,194 @@ def _search(self, trees, to_plays):
...
+            # min_value_score is not defined if none of the children is expanded.
+            # set it to 0
+            min_value_score[min_value_score == MAXIMUM_FLOAT_VALUE] = 0.
+            value_score = torch.where(unexpanded, min_value_score, value_score)
+        elif self._unexpanded_value_score == 'mean':
+            value_score[unexpanded] = 0.
+            n = (~unexpanded).sum(dim=1, keepdim=True) + 1e-30
+            mean_value_score = value_score.sum(dim=1, keepdim=True) / n
+            value_score = torch.where(unexpanded, mean_value_score,
+                                      value_score)
```
Comment (hnyu) [inline alf/algorithms/mcts_algorithm.py:521]: raise error if else

File: alf/algorithms/mcts_algorithm.py [hunk 5]
```diff
@@ -436,40 +586,129 @@ def _expand_node(
...
+
+
+def calculate_exploration_policy(value, prior, c, tol=1e-6):
+    r"""Calculate exploration policy.
+
+    The policy is based on
+    `Grill et. al. Monte-Carlo tree search as regularized policy optimization
+    <https://arxiv.org/abs/2007.12509>`_
+
+    Notation:
```
Comment (hnyu) [inline alf/algorithms/mcts_algorithm.py:633]: format broken here

File: alf/algorithms/mcts_algorithm.py [hunk 6]
```diff
@@ -536,9 +775,9 @@ def visit_softmax_temperature(num_moves):
         t = torch.ones_like(num_moves, dtype=torch.float32)
         # paper pseudocode uses 0.0
         # Current code does not support 0.0, so use a small value, which should
-        # not make any difference since a difference of 1 for visit_count translates
-        # to exp(1/0.0001) probability ratio.
-        t[num_moves >= 30] = 0.0001
+        # not make any difference since a difference of 1e-3 for visit probability
+        # to about exp(1e-3*1e10) probability ratio.
```
Comment (hnyu) [inline alf/algorithms/mcts_algorithm.py]: translates to ...

File: alf/algorithms/mcts_models.py [hunk 7]
```diff
@@ -67,13 +68,25 @@
     ])
 
 
+def _entropy(events):
```
Comment (hnyu) [inline alf/algorithms/mcts_models.py]: Should be an internal function

File: alf/algorithms/mcts_models.py [hunk 8]
```diff
@@ -195,45 +245,34 @@ def create_simple_dynamics_net(input_tensor_spec):
...
-                 debug_summaries=False,
-                 name="SimpleMCTSModel"):
+                 trunk_net_ctor,
+                 initial_game_over_bias=0.0):
         """
         Args:
-            initial_game_over_bias (float): initial bias for predicting the
+            observation_spec (TensorSpec): describing the observation.
+            action_spec (BoundedTensorSpec): describing the action.
+            trunc_net_ctor (Callable): called as ``trunk_net_ctor(input_tensor_spec=observation_spec)``
```
Comment (hnyu) [inline alf/algorithms/mcts_models.py]: trunc_net_ctor -> trunk_net_ctor

File: alf/algorithms/muzero_algorithm.py [hunk 9]
```diff
@@ -189,68 +271,106 @@ def preprocess_experience(self, experience: Experience):
...
+                    candidate_actions[r] = r_candidate_actions
+                candidate_action_policy[r] = r_candidate_action_policy
+                values[r] = r_values
+
+            game_overs = ()
+            if self._train_game_over_function or self._train_reward_function:
+                game_overs = positions == episode_end_positions
+                discount = replay_buffer.get_field('discount', env_ids,
+                                                   positions)
+                # In the case of discount == 0, the game over may not always be correct
```
Comment (hnyu) [inline alf/algorithms/muzero_algorithm.py]: Should it be "In the case of discount !=0"?

File: alf/algorithms/muzero_algorithm.py [hunk 10]
```diff
@@ -273,7 +393,176 @@ def _calc_monte_carlo_return(self, replay_buffer, env_ids, positions,
...
+            # It also needs exp.batch_info and exp.replay_buffer.
+            exp = alf.nest.map_structure(lambda x: x.unsqueeze(1), exp)
+            exp = exp._replace(
+                batch_info=BatchInfo(flat_env_ids, flat_positions),
+                replay_buffer=replay_buffer)
+            exp = self._data_transformer.transform_experience(exp)
+            exp = exp._replace(batch_info=(), replay_buffer=())
+            exp = alf.nest.map_structure(lambda x: x.squeeze(1), exp)
+
+        def _split1(x):
```
Comment (hnyu) [inline alf/algorithms/muzero_algorithm.py:457]: Can use torch.split?

File: alf/algorithms/muzero_algorithm.py [hunk 11]
```diff
@@ -273,7 +393,176 @@ def _calc_monte_carlo_return(self, replay_buffer, env_ids, positions,
...
+                                 mcts_state_field))
+
+        if len(result) == 1:
+            result = result[0]
+        else:
+            result = alf.nest.map_structure(
+                lambda *tensors: torch.cat(tensors), *result)
+        return convert_device(result)
+
+    def _reanalyze1(self, replay_buffer: ReplayBuffer, env_ids, positions,
```
Comment (hnyu) [inline alf/algorithms/muzero_algorithm.py]: This function seems too long to be read easily

File: alf/algorithms/muzero_algorithm_test.py [hunk 12]
```diff
@@ -0,0 +1,745 @@
...
+            batch_info=BatchInfo(env_ids=env_ids, positions=positions),
+            rollout_info_field='rollout_info')
+        processed_experience = muzero.preprocess_experience(experience)
+        import pprint
+        pprint.pprint(processed_experience.rollout_info)
+        alf.nest.map_structure(lambda x, y: self.assertEqual(x, y),
+                               processed_experience.rollout_info, expected)
+
+    def get_exptected_info(self, sparse_reward):
+        # yapf: disable
```
Comment (hnyu) [inline alf/algorithms/muzero_algorithm_test.py:219]: How did you calculate this expected info?

File: alf/algorithms/muzero_algorithm_test.py [hunk 13]
```diff
@@ -0,0 +1,745 @@
...
+                [[ 0.7500,  1.5000,  3.0000,  6.0000, 12.0000]],
+                [[ 1.5000,  3.0000,  6.0000, 12.0000, 12.0000]],
+                [[ 3.0000,  6.0000, 12.0000, 12.0000, 12.0000]],
+                [[ 6.0000, 12.0000, 12.0000, 12.0000, 12.0000]],
+                [[12.0000, 12.0000, 12.0000, 12.0000, 12.0000]],
+                [[12.0000, 12.0000, 12.0000, 12.0000, 12.0000]],
+                [[ 3.2500,  3.2500,  3.2500,  3.2500,  3.2500]]])))
+        self._test_preprocess_experience(train_reward_function=False, td_steps=-1, reanalyze_ratio=0., expected=expected)
+
+    def test5(self):
```
Comment (hnyu) [inline alf/algorithms/muzero_algorithm_test.py]: suggest changing the name of "testN" to more informative names

File: alf/examples/muzero_go.gin [hunk 14]
```diff
@@ -0,0 +1,85 @@
+include 'muzero.gin'
+
+import alf.environments.suite_go
+import alf.examples.muzero_go
+import alf.utils.schedulers
+
+WIDTH=6
+HEIGHT=6
```
Comment (hnyu) [inline alf/examples/muzero_go.gin]: This WIDTH and HEIGHT are for go 6x6

File: alf/utils/tensor_utils.py [hunk 15]
```diff
@@ -103,7 +103,7 @@ def _var(x):
         return var
 
     vary = _var(y)
-    r = 1 - _var(y - ypred) / (vary + 1e-30)
+    r = torch.where(vary == 0, vary, 1 - _var(y - ypred) / (vary + 1e-30))
```
Comment (hnyu) [inline alf/utils/tensor_utils.py:106]: Why setting the ratio to 0 if vary==0?

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/algorithms/data_transformer.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/data_transformer.py]: All DataTransformers are constructed in the same way by policy_trainer. So observation_spec is needed

File: alf/algorithms/mcts_algorithm.py [hunk 16]
```diff
@@ -181,11 +190,28 @@ def __init__(
             is_two_player_game: bool,
             visit_softmax_temperature_fn: Callable,
             known_value_bounds=None,
+            unexpanded_value_score=0.5,
+            act_with_exploratin_policy=False,
```
Comment (Haichao-Zhang) [inline alf/algorithms/mcts_algorithm.py]: exploratin_policy -> exploration_policy

File: alf/algorithms/mcts_algorithm.py [hunk 17]
```diff
@@ -181,11 +190,28 @@ def __init__(
...
+                1. valid_action_mask: a bool Tensor to indicate which actions are
+                    allowed. It will be used to mask out invalid actions. If not
+                    provided, all possible actions are considered.
+                2. steps: int32 Tensor to indicate the number of steps since the
+                    beginning of the game. If not provided, an internal counter
+                    will be used. However, this internal count will not be
+                    correct if the algorithm is used to play against human because
+                    it is not used to generate all the moves of both players.
+                3. to_play: int8 Tensor whose elements are 0 or 1 to indicate who
+                    is the player to take the action. If not provided, stesp % 2
```
Comment (Haichao-Zhang) [inline alf/algorithms/mcts_algorithm.py]: stesp -> steps

File: alf/algorithms/mcts_algorithm.py (hunk 16 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: fixed

File: alf/algorithms/mcts_algorithm.py (hunk 17 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: fixed

File: alf/algorithms/mcts_algorithm.py (hunk 3 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: added

File: alf/algorithms/mcts_algorithm.py (hunk 4 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py:521]: It is already checked at line 258.

File: alf/algorithms/mcts_algorithm.py (hunk 5 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py:633]: fixed

File: alf/algorithms/mcts_algorithm.py (hunk 6 above)
Comment (emailweixu) [reply alf/algorithms/mcts_algorithm.py]: updated

File: alf/algorithms/mcts_models.py (hunk 7 above)
Comment (emailweixu) [reply alf/algorithms/mcts_models.py]: updated.

File: alf/algorithms/mcts_models.py (hunk 8 above)
Comment (emailweixu) [reply alf/algorithms/mcts_models.py]: fixed

File: alf/algorithms/muzero_algorithm.py (hunk 9 above)
Comment (emailweixu) [reply alf/algorithms/muzero_algorithm.py]: Good catch. Updated.

File: alf/algorithms/muzero_algorithm.py (hunk 10 above)
Comment (emailweixu) [reply alf/algorithms/muzero_algorithm.py:457]: even with torch.split, we still need too more map_structure to separate the too splits into separate strucutre.

File: alf/algorithms/muzero_algorithm_test.py (hunk 12 above)
Comment (emailweixu) [reply alf/algorithms/muzero_algorithm_test.py:219]: By printing out the results and examining them whether they are correct.

File: alf/algorithms/muzero_algorithm_test.py (hunk 13 above)
Comment (emailweixu) [reply alf/algorithms/muzero_algorithm_test.py]: updated

File: alf/examples/muzero_go.gin (hunk 14 above)
Comment (emailweixu) [reply alf/examples/muzero_go.gin]: updated

File: alf/utils/tensor_utils.py (hunk 15 above)
Comment (emailweixu) [reply alf/utils/tensor_utils.py:106]: So that it doesn't blow up the tensorboard graph.

File: alf/algorithms/muzero_algorithm.py (hunk 11 above)
Comment (emailweixu) [reply alf/algorithms/muzero_algorithm.py]: refactored.

**Review by hnyu — DISMISSED** [review]

File: alf/algorithms/data_transformer.py (hunk 2 above)
Comment (hnyu) [reply alf/algorithms/data_transformer.py]: Docstring is missing for this class. Also, perhaps it's better to add an update_mode so that it's possible to update the normalizer during training.

File: .ci-cd/build.sh (hunk 1 above)
Comment (emailweixu) [reply .ci-cd/build.sh]: "export"  will only affect this process. So it doesn't matter.

File: alf/algorithms/data_transformer.py (hunk 2 above)
Comment (emailweixu) [reply alf/algorithms/data_transformer.py]: updated

**Review by hnyu — APPROVED** [review]
